_posts/2017-11-22-macros.md

```diff
@@ -0,0 +1,291 @@
+related problems for the past couple years as part of his research on the
+intersection of database systems, programming languages and large-scala data
+analysis.
+
+## Squid: reusable and type-safe quasiquotes
+
+[Squid] is a metaprogramming framework that facilitates the type-safe
+manipulation of Scala programs through quasiquotes.
+Unlike scala-reflect quasiquotes, Squid quasiquotes are statically guaranteed
+to be well-typed, well-scoped and hygienic.
+Another important attribute of Squid quasiquotes is that they are reusable.
+This means it's possible to write generic utility functions with Squid
+quasiquotes, which is a common limitation of alternative statically typed
+quasiquote systems.
+The properties of Squid quasiquotes make them ideal for robust and safe
+transformation of trees, the exact problem we struggled with in our prototype
+macro system.
+
+Squid quasiquotes statically track the context/scope of a quasiquote in its type to
+prevent references to unbound identifiers.
+For example, `code"(x: Int) + 1"` would be rejected since `x` is undefined.
+Instead, quasiquotes must explicitly annotate free variables
+using question marks `code"(?x: Int) + 1)` and free variables are then tracked in
+the second parameter of quasiquote's type `Code[Int, {x: Int}]`.
+
+To learn more about Squid works, I highly recommend reading their upcoming POPL
+2018 paper
+["Unifying Analytic and Statically-Typed Quasiquotes"](https://infoscience.epfl.ch/record/232427).
+Here is just one interesting table from the paper that I want to include here
+
+![Squid quasiquotes]({{ site.baseurl }}/resources/img/squid-quasiquotes.png)
+
+There are still many open question on how to incorporate Squid with the new macros.
+For example,
+
+- Squid quasiquotes eagerly convert compiler trees to Squid's IR. This may
+  introduce problems for Scala syntax like patterns and important
+  metadata may get lost like positions/attachments.
+- Squid currently does not support the ability to define methods or classes in
+  quasiquotes.
+
+I believe these problems are solvable, and we are discussing with
+Lionel and colleagues from and his lab about potential collaboration.
+If successful, the new Scala macros may have potential to push the
+state-of-the-art in metaprogramming.
+
+## Conclusion
+
+There remain many open challenges in the design of the new macros.
+I believe there is a lot for us to learn from related research in this field.
+Our next steps is to continue the work in the [scalacenter/macros] repository,
+which implements the building blocks that are required to support higher-level
+frameworks such as Squid.
+These building blocks include:
+- a rich compile-time reflection API to query for properties of types, symbols
+  and trees.
+- a standard syntax to declare and implement def macros, which will be similar
+  but distinct from the syntax to define scala-reflect def macros.
+
+If you are interested to join the effort, don't hesitate to get in touch!
```

Comment:
Say where people should go to get in touch with you 🙂

Reply:
Done 😄